### PR TITLE
SVG Circuit Representations

### PR DESCRIPTION
**Pull request by mpharrigan**
I opened #2313 in the hopes that someone else would do it and I would reap the benefits. Unfortunately, I ended up nerd sniping myself and threw this together last night.

It's rough-and-ready but I think looks pretty serviceable in the end.

image

image

### CODE REVIEW

**Comment by mpharrigan** [comment]
This is very-much a "minimum viable" version of this. I'd advocate for merging into contrib and someone else can pick it up and make it more legit. I'm not going to do any more work for it

File: cirq/contrib/svg/svg.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    import cirq
+
+
+def _get_text_width(t: str):
+    tp = matplotlib.textpath.TextPath((0, 0), t, size=14, prop='Arial')
+    bb = tp.get_extents()
+    return bb.width + 10
+
+
+def _rect(x, y, boxwidth, boxheight, fill='white', strokewidth=1):
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: Type annotations.

File: cirq/contrib/svg/svg.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    return f'<rect x="{x}" y="{y}" width="{boxwidth}" height="{boxheight}" ' \
+           f'stroke="black" fill="{fill}" stroke-width="{strokewidth}" />'
+
+
+def _text(tx, ty, text, fontsize=14):
+    return f'<text x="{tx}" y="{ty}" dominant-baseline="middle" ' \
+           f'text-anchor="middle" font-size="{fontsize}px">{text}</text>'
+
+
+def _fit_horizontal(tdd, ref_boxwidth, col_padding):
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: Type annotations.

File: cirq/contrib/svg/svg.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        col_widths[i] += padding
+
+    col_starts = [0]
+    for i in range(1, max_xi + 3):
+        col_starts.append(col_starts[i - 1] + col_widths[i - 1])
+
+    return col_starts, col_widths
+
+
+def tdd_to_svg(
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: Type annotations.

File: cirq/contrib/svg/svg.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    return t
+
+
+class SVGCircuit:
+
+    def __init__(self, circuit):
+        # coverage: ignore
+        self.circuit = circuit
+
+    def _repr_svg_(self):
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: Type annotations.

File: cirq/contrib/svg/svg.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        # coverage: ignore
+        tdd = self.circuit.to_text_diagram_drawer(transpose=False)
+        return tdd_to_svg(tdd)
+
+
+def svg_circuit(circuit):
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: Type annotations.

File: cirq/contrib/svg/svg.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            continue
+
+        t += _rect(boxx, boxy, boxwidth, boxheight)
+        t += _text(x, y, v.text, fontsize=14 if len(v.text) > 1 else 18)
+
+    t += '</svg>'
+    return t
+
+
+class SVGCircuit:
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: What is this class needed for?

Is it supposed to be a utility for writing the result to files?

Comment (mpharrigan) [reply cirq/contrib/svg/svg.py]: To get it to show up in jupyter notebook you need to wrap your circuit with `SVGCircuit`, see the example notebook

Comment (mpharrigan) [reply cirq/contrib/svg/svg.py]: We can certainly add `_repr_svg_` to `cirq.Circuit` but I don't think this code is mature enough to supplant the text diagram by default

**Comment by mpharrigan** [comment]
PTAL

File: cirq/contrib/svg/svg.py (hunk 6 above)
Comment (Strilanc) [reply cirq/contrib/svg/svg.py]: Oh I see. Add a class docstring explaining that this is intended to be consumed by a Jupyter notebook in that way.

File: cirq/contrib/svg/svg.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        # coverage: ignore
+        self.circuit = circuit
+
+    def _repr_svg_(self) -> str:
+        # coverage: ignore
+        tdd = self.circuit.to_text_diagram_drawer(transpose=False)
+        return tdd_to_svg(tdd)
+
+
+def svg_circuit(circuit: 'cirq.Circuit') -> str:
```
Comment (Strilanc) [inline cirq/contrib/svg/svg.py]: This doesn't return an `SvgCircuit`, which is a bit confusing. Maybe `circuit_to_svg`?

Comment (mpharrigan) [reply cirq/contrib/svg/svg.py]: done

File: cirq/contrib/svg/svg.py (hunk 6 above)
Comment (mpharrigan) [reply cirq/contrib/svg/svg.py]: added

**Comment by Strilanc** [comment]
Windows build flaked on the choco install python3.7 line. Restarted it.

**Pull request merged**